stroyprombeton/context.py

```diff
@@ -106,36 +106,25 @@ def category(self):
     def tags(self) -> context.Tags:
         return context.Tags(stb_models.Tag.objects.all())
 
-    def slice_products(
-        self, options: stb_models.OptionQuerySet
-    ) -> context.products.PaginatedProducts:
-        """
-        We have to use separated variable/method for pagination.
-
-        Because paginated QuerySet can not used as QuerySet.
-        It's not the most strong place of Django ORM, of course.
-        :return: ProductsContext with paginated QuerySet inside
-        """
-        return context.products.PaginatedProducts(
-            products=options,
+    def context(self) -> typing.ContextDict:
+        tags = FilteredTags(stb_models.Tag.objects.all(), self.request_data)
+        options = FilteredOptions(self.category, tags.qs(), self.request_data)
+
+        # @todo #514:60m  Create PaginatedOptions class.
+        #  Without code doubling between the new class
+        #  and `context.products.PaginatedProducts` one.
+        sliced_options = context.products.PaginatedProducts(
+            products=options.qs(),
             url=self.request_data.request.path,
             page_number=self.request_data.pagination_page_number,
             per_page=self.request_data.pagination_per_page,
         )
 
-    def filter_options(self, tags):
-        return FilteredOptions(self.category, tags.qs(), self.request_data).qs()
-
-    def context(self) -> typing.ContextDict:
-        tags = FilteredTags(stb_models.Tag.objects.all(), self.request_data)
-        options_qs = self.filter_options(tags)
-        sliced_options = self.slice_products(options_qs)
-
         images = context.products.ProductImages(
             sliced_options.products, Image.objects.all()
         )
         grouped_tags = context.tags.GroupedTags(
-            tags=TagsByOptions(self.tags, options_qs)
+            tags=TagsByOptions(self.tags, options.qs())
```

Comment:
Let's pass whole `Context` instance to ctor of decorators to avoid such issues. If you create a todo for this it would be nice

Reply:
@artemiy312 , already created it below
https://github.com/fidals/stroyprombeton/pull/516/files/c15d80750e853035f048b534466478f84cb2240a#diff-f5058f0d4ecd89be6a462ef5ed432810R263